src/tools/fuzzing/heap-types.cpp

```diff
@@ -0,0 +1,499 @@
+      switch (type.getBasic()) {
+        case HeapType::func:
+          return pickSubFunc();
+        case HeapType::ext:
+          return HeapType::ext;
+        case HeapType::any:
+          return pickSubAny();
+        case HeapType::eq:
+          return pickSubEq();
+        case HeapType::i31:
+          return HeapType::i31;
+        case HeapType::data:
+          return pickSubData();
+      }
+      WASM_UNREACHABLE("unexpected kind");
+    }
+  }
+
+  // TODO: Make this part of the wasm-type.h API
+  struct Ref {
+    HeapType type;
+    Nullability nullability;
+  };
+
+  Ref generateSubRef(Ref super) {
+    auto nullability = super.nullability == NonNullable
+                         ? NonNullable
+                         : rand.oneIn(2) ? Nullable : NonNullable;
+    return {pickSubHeapType(super.type), nullability};
+  }
+
+  Rtt generateSubRtt(Rtt super) {
+    auto depth = super.hasDepth()
+                   ? super.depth
+                   : rand.oneIn(2) ? Rtt::NoDepth : rand.upTo(MAX_RTT_DEPTH);
+    return {depth, super.heapType};
+  }
+
+  Type generateSubtype(Type type) {
+    if (type.isBasic()) {
+      // We do not construct types with basic reference types (we go through the
+      // TypeBuilder for those instead), so this must be a non-reference basic
+      // type, which means it has no other subtypes.
+      return type;
+    } else if (type.isRef()) {
+      auto ref = generateSubRef({type.getHeapType(), type.getNullability()});
+      return builder.getTempRefType(ref.type, ref.nullability);
+    } else if (type.isRtt()) {
+      auto rtt = generateSubRtt(type.getRtt());
+      return builder.getTempRttType(rtt);
+    } else {
+      WASM_UNREACHABLE("unexpected type kind");
+    }
+  }
+
+  Signature generateSubSignature(Signature super) {
+    // TODO: Update this once we support nontrivial function subtyping.
+    return super;
+  }
+
+  Field generateSubField(Field super) {
+    if (super.mutable_ == Mutable) {
+      // Only immutable fields support subtyping.
+      return super;
+    }
+    if (super.isPacked()) {
+      // No other subtypes of i8 or i16.
+      return super;
+    }
+    return {generateSubtype(super.type), Immutable};
+  }
+
+  Struct generateSubStruct(const Struct& super) {
+    if (rand.oneIn(2)) {
+      return super;
+    }
+    std::vector<Field> fields;
+    // Depth subtyping
+    for (auto field : super.fields) {
+      fields.push_back(generateSubField(field));
+    }
+    // Width subtyping
+    Index extra = rand.upTo(MAX_STRUCT_SIZE + 1 - fields.size());
+    for (Index i = 0; i < extra; ++i) {
+      fields.push_back(generateField());
+    }
+    return {fields};
+  }
+
+  Array generateSubArray(Array super) {
+    if (rand.oneIn(2)) {
+      return super;
+    }
+    return {generateSubField(super.element)};
+  }
+
+  HeapTypeKind generateHeapTypeKind() {
+    // Building basic heap types is only allowed in equirecursive mode.
+    // TODO: Relax this.
```

Comment:
This is an artificial limitation of the current implementation. When I wrote the nominal implementation, I cut some corners and assumed / asserted that none of the TypeBuilder slots would be basic types. This is fine for normal use because the only time wasm-opt has to build a basic type is in equirecursive LUB calculations. Nominal typing should never need to build a basic type, but it turns out that having this difference between the type systems makes it harder to generalize code to work with both of them.